src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/AbstractRazorDelegatingEndpoint.cs

```diff
@@ -0,0 +1,140 @@
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using MediatR;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common.Extensions;
+using Microsoft.AspNetCore.Razor.LanguageServer.Extensions;
+using Microsoft.AspNetCore.Razor.LanguageServer.Hover;
+using Microsoft.AspNetCore.Razor.LanguageServer.Protocol;
+using Microsoft.CodeAnalysis.Razor.Workspaces;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.Extensions.Logging;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using OmniSharp.Extensions.JsonRpc;
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer
+{
+    internal abstract class AbstractRazorDelegatingEndpoint<TRequest, TResponse, TDelegatedParams> : IJsonRpcRequestHandler<TRequest, TResponse?>
+        where TRequest : TextDocumentPositionParams, IRequest<TResponse>
+        where TResponse : class
+        where TDelegatedParams : class
+    {
+        private readonly DocumentContextFactory _documentContextFactory;
+        private readonly LanguageServerFeatureOptions _languageServerFeatureOptions;
+        private readonly RazorDocumentMappingService _documentMappingService;
+        private readonly ClientNotifierServiceBase _languageServer;
+        private readonly ILogger<RazorHoverEndpoint> _logger;
+
+        protected AbstractRazorDelegatingEndpoint(
+            DocumentContextFactory documentContextFactory,
+            LanguageServerFeatureOptions languageServerFeatureOptions,
+            RazorDocumentMappingService documentMappingService,
+            ClientNotifierServiceBase languageServer,
+            ILoggerFactory loggerFactory)
+        {
+            _documentContextFactory = documentContextFactory ?? throw new ArgumentNullException(nameof(documentContextFactory));
+            _languageServerFeatureOptions = languageServerFeatureOptions ?? throw new ArgumentNullException(nameof(languageServerFeatureOptions));
+            _documentMappingService = documentMappingService ?? throw new ArgumentNullException(nameof(documentMappingService));
+            _languageServer = languageServer ?? throw new ArgumentNullException(nameof(languageServer));
+            _logger = loggerFactory?.CreateLogger<RazorHoverEndpoint>() ?? throw new ArgumentNullException(nameof(loggerFactory));
+        }
+
+        /// <summary>
+        /// Override to provide fallback behavior for when single server is not supported
+        /// </summary>
+        protected abstract Task<TResponse?> HandleWithoutSingleServerAsync(TRequest request, DocumentContext documentContext, RazorCodeDocument codeDocument, SourceText sourceText, CancellationToken cancellationToken);
```

Comment:
oo interesting approach. When we remove the option to not be single server do you expect this to be an easy removal?

Reply:
I haven't read further than this, so this might be answered later, but how would this be different to what is in HandleInRazorAsnyc?

I would have thought that in non-single server mode, we'd just never call the HandleDelegated method.

Reply:
I have now read the whole PR, and I come to you from the future to say that I agree with past Dave.. we should just have HandleRazor and Delegate, and in non-single-server this class should just not call Delegate. I also agree with past Dave about HandleAsync not being protected.